File: cakk-api/src/main/java/com/cakk/api/listener/ErrorAlertEventListener.java
package com.cakk.api.listener;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.scheduling.annotation.Async;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

import com.cakk.api.annotation.ApplicationEventListener;
import com.cakk.api.dto.event.ErrorAlertEvent;
import com.cakk.api.mapper.EventMapper;
import com.cakk.external.extractor.MessageExtractor;
import com.cakk.external.sender.MessageSender;
import com.cakk.external.template.MessageTemplate;
import com.cakk.external.vo.ErrorAlertMessage;

@ApplicationEventListener
public class ErrorAlertEventListener {

	private final MessageTemplate messageTemplate;
	private final MessageExtractor messageExtractor;
	private final MessageSender messageSender;

	public ErrorAlertEventListener(
		MessageTemplate messageTemplate,
		@Qualifier("errorAlertMessageExtractor") MessageExtractor messageExtractor,
		@Qualifier("slackMessageSender") MessageSender messageSender
	) {
		this.messageTemplate = messageTemplate;
		this.messageExtractor = messageExtractor;
		this.messageSender = messageSender;
	}

	@Async
	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
	public void sendMessageToSlack(ErrorAlertEvent errorAlertEvent) {
		ErrorAlertMessage certificationMessage = EventMapper.supplyErrorAlertMessageBy(errorAlertEvent);
		messageTemplate.sendMessage(certificationMessage, messageExtractor, messageSender);
	}
}

File: cakk-core/src/main/kotlin/com/cakk/core/listener/ErrorAlertEventListener.kt
package com.cakk.core.listener

import org.springframework.scheduling.annotation.Async
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

import net.gpedro.integrations.slack.SlackMessage

import com.cakk.core.annotation.ApplicationEventListener
import com.cakk.core.dto.event.ErrorAlertEvent
import com.cakk.core.mapper.supplyErrorAlertMessageBy
import com.cakk.external.extractor.MessageExtractor
import com.cakk.external.sender.MessageSender
import com.cakk.external.template.MessageTemplate
import com.cakk.external.vo.message.ErrorAlertMessage

@ApplicationEventListener
class ErrorAlertEventListener(
	private val messageTemplate: MessageTemplate,
	private val messageExtractor: MessageExtractor<ErrorAlertMessage, SlackMessage>,
	private val messageSender: MessageSender<SlackMessage>
) {

    @Async
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    fun sendMessageToSlack(errorAlertEvent: ErrorAlertEvent) {
        val certificationMessage = supplyErrorAlertMessageBy(errorAlertEvent)
        messageTemplate.sendMessage(certificationMessage, messageExtractor, messageSender)
    }
}
